Speed up `deleted_reason` method

apainintheneck opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

When playing around with brew a few days ago I realized that one of the reasons that `brew search` and `brew info` can be slow sometimes for me is because of the following `git log` command in `MissingFormula.deleted_reason`.

https://github.com/Homebrew/brew/blob/b683bebd24b5003b47e722116eb2a0989e1e7111/Library/Homebrew/missing_formula.rb#L157-L159

```sh
$ time brew info unknown_formula
Error: No available formula with the name "unknown_formula".
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.

________________________________________________________
Executed in    9.96 secs    fish           external
   usr time    8.72 secs   13.84 millis    8.71 secs
   sys time    1.03 secs    2.77 millis    1.02 secs

$ pwd
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
$ time git log --since='1 month ago'\
            --diff-filter=D --max-count=1\
            --format=%H\\\\n%h\\\\n%B -- Formula/unknown_formula.rb

________________________________________________________
Executed in    7.38 secs    fish           external
   usr time    7.31 secs  209.00 micros    7.31 secs
   sys time    0.07 secs  603.00 micros    0.07 secs
```

After investigating a little more I found out that we default to searching the core tap when a formula is not found by name in any of the taps. This `git log` command searches linearly through the last months worth of commits which is thousands of commits on the core tap. I decided to limit this optimization to the core tap because I don't think that any other taps are active enough for this to be a performance issue. Does anyone know if there are any other taps of a similar size out there?

```sh
$ pwd
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
$ git log --since='1 month ago' --pretty=oneline | wc -l
   10213
$ git log --before='1 month ago' --after='2 month ago'\
                                         --pretty=oneline | wc -l
    5195
$ git log --before='2 month ago' --after='3 month ago'\
                                         --pretty=oneline | wc -l
    4852
```

To speed this up I just check to see if the formula existed in a commit from a month ago before doing the entire `git log` command. The tradeoff here is that we will miss any formulas that were introduced less than a month ago and then removed. That seems unlikely but it is a tradeoff. I tend to think it's more likely that people will search for something that doesn't exist or make a typo. Also, we don't remove that many formulae every month. In the past month, the only one we removed was `ksh` but that now points to `ksh93` because `deleted_reason` never actually gets called when searching for deprecated formulae.

```sh
$ pwd
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
$ git log --diff-filter=D --since='1 year ago' --pretty=oneline | wc -l
     101
```

We removed 101 formulae over the past year but it's very inconsistent from month to month (as I already mentioned we've only removed one in the past month).

Here are some before and after benchmarks.

### Before
```sh
$ time brew info unknown_formula
Error: No available formula with the name "unknown_formula".
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.

________________________________________________________
Executed in    9.68 secs    fish           external
   usr time    8.48 secs  196.00 micros    8.48 secs
   sys time    1.00 secs  515.00 micros    1.00 secs

$ time brew info jjrnl
Error: No available formula with the name "jjrnl". Did you mean jrnl?
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.

________________________________________________________
Executed in    6.10 secs    fish           external
   usr time    4.96 secs  191.00 micros    4.96 secs
   sys time    1.00 secs  584.00 micros    1.00 secs

$ time brew search unknown_formula
Error: No formulae or casks found for "unknown_formula".

________________________________________________________
Executed in   15.13 secs    fish           external
   usr time   10.76 secs  207.00 micros   10.76 secs
   sys time    2.24 secs  642.00 micros    2.23 secs

$ time brew search jjrnl
==> Formulae
jrnl ✔

________________________________________________________
Executed in    9.80 secs    fish           external
   usr time    7.16 secs  198.00 micros    7.16 secs
   sys time    1.85 secs  551.00 micros    1.85 secs
```

### After
```sh
$ time brew info unknown_formula
Error: No available formula with the name "unknown_formula".
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.

________________________________________________________
Executed in    2.39 secs    fish           external
   usr time    1.25 secs  190.00 micros    1.25 secs
   sys time    0.99 secs  563.00 micros    0.99 secs

$ time brew info jjrnl
Error: No available formula with the name "jjrnl". Did you mean jrnl?
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.

________________________________________________________
Executed in    2.38 secs    fish           external
   usr time    1.22 secs  198.00 micros    1.22 secs
   sys time    1.00 secs  529.00 micros    1.00 secs

$ time brew search unknown_formula
Error: No formulae or casks found for "unknown_formula".

________________________________________________________
Executed in    6.20 secs    fish           external
   usr time    3.45 secs  199.00 micros    3.45 secs
   sys time    1.84 secs  611.00 micros    1.84 secs

$ time brew search jjrnl
==> Formulae
jrnl ✔

________________________________________________________
Executed in    6.18 secs    fish           external
   usr time    3.44 secs  202.00 micros    3.44 secs
   sys time    1.87 secs  617.00 micros    1.87 secs
```

Is this a worthwhile tradeoff to speed up `brew search` and `brew info`? Also, I'd be interested to see what the benchmarks look like on newer macs.

apainintheneck commented:
I thought I'd add my git version too since that could also be a factor in the speed of the `git log` command.

```sh
$ git version
git version 2.24.3 (Apple Git-128)
```

apainintheneck commented:
Okay, I reran the benchmarks with my work laptop and there still seems to be a significant benefit. The big difference seems to be that the newer version of git on the M1 Macbook runs the `git log` command about twice as fast as on my old computer.

```sh
$ pwd
/opt/homebrew/Library/Taps/homebrew/homebrew-core/o/h/L/T/h/homebrew-core
$ time git log --since='1 month ago'\
    --diff-filter=D --max-count=1\
    --format=%H\\\\n%h\\\\n%B -- Formula/unknown_formula.rb

________________________________________________________
Executed in    3.74 secs    fish           external
   usr time    3.65 secs    0.33 millis    3.65 secs
   sys time    0.05 secs    2.31 millis    0.04 secs
```

It seems like this change would still save a few seconds when using the `brew info` and `brew search` commands.

|cmd|pc|before|after|
|--|--|--|--|
|`brew info jjrnl`|old iMac|6.10 secs|2.38 secs|
|`brew info jjrnl`|M1 Macbook|2.73 secs| 894.83 millis|
|`brew info unknown_formula`|old iMac|9.68 secs|2.39 secs|
|`brew info unknown_formula`|M1 Macbook|4.46 secs|2.56 secs|
|`brew search jjrnl`|old iMac|9.80 secs|6.18 secs|
|`brew search jjrnl`|M1 Macbook|7.91 secs|2.06 secs|
|`brew search unknown_formula`|old iMac|15.30 secs|6.20 secs|
|`brew search unknown_formula`|M1 Macbook|6.29 secs| 2.95 secs|

Legend:
<details>
<summary>Old iMac: Brew Config</summary>

```
HOMEBREW_VERSION: 3.6.10-9-ge1b2f68
ORIGIN: https://github.com/Homebrew/brew
HEAD: e1b2f68a6d0c151254f2686b50ec661b973f83b6
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 24255ed4817a3b54f91a1aa893140c8017a67b18
Core tap last commit: 2 days ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_AUTOREMOVE: set
HOMEBREW_CASK_OPTS: []
HOMEBREW_DEVELOPER: set
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.D3193WXYad/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: quad-core 64-bit ivybridge
Clang: 12.0.0 build 1200
Git: 2.24.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.0.32.29
Xcode: 12.4
```

</details>
<details>
<summary>M1 Macbook: Brew Config</summary>

```
HOMEBREW_VERSION: 3.6.10-11-gb683beb
ORIGIN: https://github.com/Homebrew/brew
HEAD: b683bebd24b5003b47e722116eb2a0989e1e7111
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 12cb55611f0884a68921be5a0badf29b2b0e3f48
Core tap last commit: 60 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.38.1 => /opt/homebrew/bin/git
Curl: 7.84.0 => /usr/bin/curl
macOS: 13.0-arm64
CLT: 14.1.0.0.1.1666437224
Xcode: N/A
Rosetta 2: false
```

</details>

MikeMcQuaid commented:
> I decided to limit this optimization to the core tap because I don't think that any other taps are active enough for this to be a performance issue. Does anyone know if there are any other taps of a similar size out there?

`homebrew-cask`?

> Is this a worthwhile tradeoff to speed up `brew search` and `brew info`?

I think so! Note `hyperfine` is the best tool for doing these sorts of benchmarks.

MikeMcQuaid commented on Library/Homebrew/missing_formula.rb:
```diff
@@ -154,6 +154,25 @@ def deleted_reason(name, silent: false)
+          # command below a performance issue when using `brew search` or
+          # `brew info` to look for a formula that doesn't exist. This first
+          # checks if the formula existed a month ago before continuing.
+          if tap.is_a? CoreTap
```
```suggestion
          if tap.core_tap?
```

MikeMcQuaid commented on Library/Homebrew/missing_formula.rb:
```diff
@@ -154,6 +154,25 @@ def deleted_reason(name, silent: false)
+          # `brew info` to look for a formula that doesn't exist. This first
+          # checks if the formula existed a month ago before continuing.
+          if tap.is_a? CoreTap
+            commit_command = "git log --before='1 month ago' --max-count=1 --format=%H"
+            month_old_commit_hash = Utils.popen_read(commit_command).chomp
```
```suggestion
            commit_command = ["git", "log", "--before='1 month ago'", "--max-count=1", "--format=%H"]
            month_old_commit_hash = Utils.popen_read(*commit_command).chomp
```

MikeMcQuaid commented on Library/Homebrew/missing_formula.rb:
```diff
@@ -154,6 +154,25 @@ def deleted_reason(name, silent: false)
+
+            # Check if the formula has been deleted in the last month
+            # by comparing the diff between now and a month ago.
+            diff_command = "git diff --diff-filter=D --name-only " \
+                           "#{month_old_commit_hash} HEAD -- #{relative_path}"
+
+            if Utils.popen_read(diff_command).blank?
```
```suggestion
            diff_command = ["git", "diff", "--diff-filter=D", "--name-only",
                            month_old_commit_hash, "HEAD", "--", relative_path]

            if Utils.popen_read(@diff_command).blank?
```

MikeMcQuaid reviewed (commented):
Nice work so far! Some thoughts:
- always nicer to use arrays of arguments when possible to avoid creating a subshell unnecessarily 
- are you sure we want to use `HEAD` here rather than e.g. `origin/HEAD` or `origin/master`?
- I think you could do this with a single command with e.g. `@{'1 month ago'}` instead of `git log` to get `month_old_commit_hash`.

Bo98 commented:
> I decided to limit this optimization to the core tap because I don't think that any other taps are active enough for this to be a performance issue. Does anyone know if there are any other taps of a similar size out there?

To this I ask: is this optimisation a (non-trivial) performance penalty for small taps? If not, might as well avoid the question of tap size and do it unconditionally?

apainintheneck commented:
> > I decided to limit this optimization to the core tap because I don't think that any other taps are active enough for this to be a performance issue. Does anyone know if there are any other taps of a similar size out there?
> 
> To this I ask: is this optimisation a (non-trivial) performance penalty for small taps? If not, might as well avoid the question of tap size and do it unconditionally?

No, it's trivial. The reason I was avoiding using it for all taps is that it is less thorough. It will miss any formulae that were added less than a month ago and then removed in the meantime; the current algorithm will still catch those. I think that probably doesn't happen that frequently but it's hard to know. It would simplify the code if we decided to do it for everything though.

apainintheneck commented:
> * I think you could do this with a single command with e.g. `git diff @{'1 month ago'}` instead of `git log` to get `month_old_commit_hash`.

I didn't know how to do that. Very interesting! It looks like the following should work.

```sh
git diff @{'1 month ago'} --diff-filter=D --name-only -- Formula/{name}.rb
```

> * are you sure we want to use HEAD here rather than e.g. origin/HEAD or origin/master?

I assume the above command uses HEAD by default. I didn't really think too hard about it to be honest. Is there any downside to doing it this way? If we decide to apply this to all taps, would we be able to count on that always being the branch name? Is that enforced anywhere?

> Note `hyperfine` is the best tool for doing these sorts of benchmarks.

I'll keep that in mind for next time.

apainintheneck commented:
There's also a part of me that wonders if there will be a gap between `--since='1 month ago'` and `@{'1 month ago'}`. Maybe it'd be better to be explicit? Check for the existence of the file at `@{'30 days ago'}` and then run the git log command if it exists with `--since='31 days ago'`.

RandomDSdevel commented:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Alternatively or in addition, would using [the cache store](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cache_store.rb) per-formula help here any?  Of course, that only works between '`brew update`' runs, which invalidate any cached Git logs (unless only new entries were added into the cache incrementally.)  That might well be too many abstractions, though?

apainintheneck commented:
I updated the shell command so that now only one is needed as recommended by @MikeMcQuaid. I also changed it so that this optimization applies to all taps as mentioned by @Bo98. I'm still not sure if that's the right idea but the code is simpler which is nice.

MikeMcQuaid commented:
> To this I ask: is this optimisation a (non-trivial) performance penalty for small taps? If not, might as well avoid the question of tap size and do it unconditionally?

I'd flip this round: unless you can demonstrate a speed-up on small taps: it's worth just special-casing homebrew-core.

Smaller taps are in general updated less often so otherwise we make this a lot less useful as a feature for the sake of consistency (if there's minimal or no performance benefit).

MikeMcQuaid commented:
> I assume the above command uses HEAD by default. I didn't really think too hard about it to be honest. Is there any downside to doing it this way? If we decide to apply this to all taps, would we be able to count on that always being the branch name? Is that enforced anywhere?

You can specify a revision here. I think it's generally best to compare `origin/master` to older commits on `origin/master` to avoid causing bugs when users have weird stuff checked out (which happens not uncommonly).

RandomDSdevel commented:
> Do you mean trying to cache the recently deleted formulae? …

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;😮‍💨  Right, doing that directly would obviously be better.  Why the thought of caching the Git logs came to mind, I have no idea; they'd still need parsing every time that way.  (Well, except that that's the expensive operation.)  Ideally, yes, the preferred approach would be to cache _after_ parsing; that'd be cheaper still, though the cost of loading the database would naturally remain.  

> …The `CacheStoreDatabase` is currently only used to hold package descriptions and linkage information so we'd have to write new code to track when formulae are deleted and when to invalidate that cache. I'm not sure that it's worth it but it's an interesting idea.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Yeah, definitely not a blocking suggestion by any means, but the thought did occur to me, so I just thought I'd share it.

Bo98 commented:
> I'd flip this round: unless you can demonstrate a speed-up on small taps: it's worth just special-casing homebrew-core.

Less conditions mean less things to test. Would we have a test for this for both core and non-core?

Of course, if there's a functional benefit to separation then it can be worth it.

apainintheneck commented:
> I'd flip this round: unless you can demonstrate a speed-up on small taps: it's worth just special-casing homebrew-core.
> 
> Smaller taps are in general updated less often so otherwise we make this a lot less useful as a feature for the sake of consistency (if there's minimal or no performance benefit).

There won't be any speed up for small taps but at the same time it won't slow them down either. The check itself is negligible performance-wise. The question is how thorough we want to be.

> You can specify a revision here. I think it's generally best to compare `origin/master` to older commits on `origin/master` to avoid causing bugs when users have weird stuff checked out (which happens not uncommonly).

If we decide to do this for all taps, I don't think it makes sense to specify the branch for the diff command since we can't assume the name of a tap's main branch. Plus, the log command doesn't specify the branch either so I'm not sure it really matters. This would only show false positives if users started manually deleting formulae from taps which I don't think we should be worried about.

MikeMcQuaid commented:
> > I'd flip this round: unless you can demonstrate a speed-up on small taps: it's worth just special-casing homebrew-core.
> 
> Less conditions mean less things to test. Would we have a test for this for both core and non-core?

User experience >>>> how we do our testing/our test coverage.

> The question is how thorough we want to be.

As thorough as we can be while maintaining good performance.

> I don't think it makes sense to specify the branch for the diff command since we can't assume the name of a tap's main branch

`origin/HEAD` should work (or can be made to work) for all taps, just FYI.

I do think for homebrew/core it's fine to assume either `origin/HEAD` or `origin/master` here.

> This would only show false positives if users started manually deleting formulae from taps which I don't think we should be worried about.

This happens.

apainintheneck commented:
@MikeMcQuaid You convinced me so I added back in the check for the core tap.

I also added some tests. Adding the tests was more difficult than I thought though. I first tried to imitate the current tests for `deleted_reason` by creating a tap manually using git and then adding and deleting a formula. I ran into all sorts of problems related to inconsistent formula paths being returned from `Formulary.path` with my pseudo core tap. Anyway, I decided just to mock the call to `Tap#core_tap?` to be sure that the new code got exercised instead of fussing around with it anymore.

apainintheneck commented:
I didn't add in the `origin/master` part to the `git diff` query mostly because I couldn't figure out how to make that play nice with the tests.

MikeMcQuaid reviewed (approved):
Looks great, thanks @apainintheneck and sorry for all the back and forth.
